ruma-api-macros/src/api/request.rs

```diff
@@ -592,6 +592,214 @@ impl Request {
+                    // ':' should only ever appear at the start of a segment
+                    assert_eq!(&format_string[start_of_segment - 1..start_of_segment], "/");
+
+                    let end_of_segment = match format_string[start_of_segment..].find('/') {
+                        Some(rel_pos) => start_of_segment + rel_pos,
+                        None => format_string.len(),
+                    };
+
+                    let path_var = Ident::new(
+                        &format_string[start_of_segment + 1..end_of_segment],
+                        Span::call_site(),
+                    );
+                    format_args.push(quote! {
+                        #percent_encoding::utf8_percent_encode(
+                            &self.#path_var.to_string(),
+                            #percent_encoding::NON_ALPHANUMERIC,
+                        )
+                    });
+                    format_string.replace_range(start_of_segment..end_of_segment, "{}");
+                }
+
+                quote! {
+                    format_args!(#format_string, #(#format_args),*)
+                }
+            };
+
+            let path_fields =
+                path_string[1..].split('/').enumerate().filter(|(_, s)| s.starts_with(':')).map(
+                    |(i, segment)| {
+                        let path_var = &segment[1..];
+                        let path_var_ident = Ident::new(path_var, Span::call_site());
+                        quote! {
+                            #path_var_ident: {
+                                let segment = path_segments[#i].as_bytes();
+                                let decoded = #ruma_api::try_deserialize!(
+                                    self,
+                                    #percent_encoding::percent_decode(segment)
+                                        .decode_utf8(),
+                                );
+
+                                #ruma_api::try_deserialize!(
+                                    self,
+                                    ::std::convert::TryFrom::try_from(&*decoded),
+                                )
+                            }
+                        }
+                    },
+                );
+
+            (format_call, quote! { #(#path_fields,)* })
+        } else {
+            (quote! { metadata.path.to_owned() }, TokenStream::new())
+        }
+    }
+
+    /// The function determines the type of query string that needs to be built
+    /// and then builds it using `ruma_serde::urlencoded::to_string`.
+    fn build_query_string(&self, ruma_api: &TokenStream) -> TokenStream {
+        let ruma_serde = quote! { #ruma_api::exports::ruma_serde };
+
+        if let Some(field) = self.query_map_field() {
+            let field_name = field.ident.as_ref().expect("expected field to have identifier");
+
+            quote!({
+            // This function exists so that the compiler will throw an
```

Comment:
Indentation is wrong here (this should be indented more).

Reply:
It might be that rustfmt will fix this automatically after the line I highlighted below is split up.

Reply:
I did ran cargo fmt and cargo clippy before submitting the the PR. seems like I will need to do it manually :+1: .